Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+    next_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
+    actions = np.zeros((number_in_batch))  # (Array instead of list)
+
+    rewards = list()  # List
+    is_terminals = list()  # List
+
+    for i in range(len(indices_chosen)): # Probably more efficient way to do this: "for index in ?(indices_chosen)?"
+        index = indices_chosen[i] # index corresponds to the iterations
+        start_states[i, :, :, :] = make_state(mem_frames, index - 1)
+        next_states[i, :, :, :] = make_state(mem_frames, index) # State given by index was arrived at by taking action given by index and reward received is give by index.
+                                                                # In contrast, index-1 labels the previous state, before the action was taken
+        actions[i] = mem_actions.recall(index)
+
+        rewards.append(mem_rewards.recall(index))
+        is_terminals.append(mem_is_terminal.recall(index))
+
+    # We should now have a full batch, which the DNN can train on
+    fit_batch_target(model, target_model, gamma, start_states, actions, rewards, next_states, is_terminals, number_in_batch)
+
+
+def q_iteration(env, model, target_model, iteration, mem_frames, mem_actions, mem_rewards, mem_is_terminal, mem_size):
+    """
+    Do one iteration of acting then learning
+    """
+
+    # Choose epsilon based on the iteration
+    epsilon = get_epsilon_for_iteration(iteration)
+
+    start_state = make_state(mem_frames, iteration)
+    # Choose the action:
+    if random.random() < epsilon:
+        action = env.action_space.sample()
+    else:
+        action = choose_best_action(model, start_state)
+    # Play one game iteration (note: according to a paper, you should actually play 4 times here)
+    new_frame, reward, is_terminal, _ = env.step(action)
+
+    reward = transform_reward(reward)  # Make reward just +1 or -1
+    new_frame = preprocess(new_frame)  # Preprocess frame before saving it
+
+    # Add to memory
+    mem_frames.append(new_frame, iteration+1)  # The frame we ended up in after the action
+    mem_actions.append(action, iteration+1)  # The action we did
+    mem_rewards.append(reward, iteration+1)  # The reward we received after doing the action
+    mem_is_terminal.append(is_terminal, iteration+1)  # Whether or not the new frame is terminal
+
+    # Make then fit a batch (gamma=0.99, num_in_batch=?)
+    number_in_batch = 32
+    make_n_fit_batch(model, target_model, 0.99, iteration, mem_size, mem_frames, mem_actions, mem_rewards, mem_is_terminal, number_in_batch)
+
+    # If DONE, reset model!
+    if is_terminal:
+        env.reset()
+
+    return action, reward, is_terminal, epsilon
+
+#----------------------------------------------------------------------------------------------------------------------#
+
+def main():
+    """
+    Train the DNN to play Breakout
+    """
+    import argparse
```

Comment:
I'd generally recommend putting all imports at the beginning of the file unless there's a particular reason not to (e.g. it's slow to import the module and the function is only sometimes called; it's an optional dependency). See https://stackoverflow.com/questions/3095071/in-python-what-happens-when-you-import-inside-of-a-function for discussion